UpdateChecker

IhateTrains opened the pull request:
This is what appears if a tag with higher version is found on startup:
![image](https://user-images.githubusercontent.com/29546927/93824569-f32d1000-fc63-11ea-9035-11be671053a3.png)

A button:
![image](https://user-images.githubusercontent.com/29546927/93824814-5d45b500-fc64-11ea-9e89-6e6ea8f0d73a.png)

With `enableUpdateChecker = false`, the messages and the button don't appear.
With `checkForUpdatesOnStartup = false`, the check is only performed on button click.

Example configuration:

> enableUpdateChecker = true
checkForUpdatesOnStartup = true
converterTagsGithubUrl = "https://api.github.com/repos/ParadoxGameConverters/CK2ToEU4/tags"
converterReleaseForumThread = "https://forum.paradoxplaza.com/forum/threads/ck2-to-eu4-converter-release-thread.1377860/"

IhateTrains commented:
Ideas for improvement?

Zemurin commented on Fronter/Source/Frames/MainFrame.cpp:
```diff
@@ -99,6 +101,17 @@ void MainFrame::OnSupportUs(wxCommandEvent& event)
 
+void MainFrame::OnCheckForUpdates(wxCommandEvent& event)
+{
+	if (isUpdateAvailable(configuration->getConverterFolder() + "/configurables/version.txt", configuration->getConverterTagsGithubUrl()))
```
does this work if there's no getConverterFolder()?
on linux I bet it'll break.

IhateTrains replied:
Made the function break when `getConverterFolder()` returns an empty string.

Zemurin commented on Fronter/Source/UpdateChecker/UpdateChecker.cpp:
```diff
@@ -0,0 +1,216 @@
+		const auto newestDotPos = newestVersion.find('.');
+		const auto versionDotPos = version.find('.');
+		
+		if (std::stoi(version) > std::stoi(newestVersion))
```
please try/catch around all stois.

Idhrendur commented on Fronter/Source/UpdateChecker/UpdateChecker.cpp:
```diff
@@ -0,0 +1,216 @@
+	size_t nameEnd = 0;
+	do
+	{
+		pos = json.find("\"name\"", nameEnd);
```
I'm torn. On the one hand, that's not how you're supposed to do json, and it's likely really fragile. On the other, it saved us a json library.

IhateTrains replied:
We could use this: https://github.com/nlohmann/json

Idhrendur commented on Fronter/Source/UpdateChecker/UpdateChecker.cpp:
```diff
@@ -0,0 +1,216 @@
+		const auto newestDotPos = newestVersion.find('.');
+		const auto versionDotPos = version.find('.');
+		
+		if (std::stoi(version) > std::stoi(newestVersion))
```
I think you meant to extract the left sides here.

IhateTrains replied:
Doesn't `stoi` do that for us? For example `stoi("1.5F")` should return 1.

Idhrendur replied:
Hmm, perhaps technically true, but take my comment as an indication that it's unclear to people reading the code.

Idhrendur commented on Fronter/Source/UpdateChecker/UpdateChecker.cpp:
```diff
@@ -0,0 +1,216 @@
+		// if parts before dot are equal, compare the rest
+		auto newestRSide = newestDotPos == std::string::npos ? "" : newestVersion.substr(newestDotPos + 1);
+		auto versionRSide = versionDotPos == std::string::npos ? "" : version.substr(versionDotPos + 1);
+		if (version.size() > newestVersion.size())
```
You should be able to replace these loops with judicious use of std::transform (see the first example [here](https://en.cppreference.com/w/cpp/algorithm/transform)) and std::lexicographical_compare

Idhrendur replied:
It might even invalidate the need to compare both parts separately.

Idhrendur reviewed (changes requested):
More straightforward than I expected, but I hadn't ever really put thought to version checking.

However, it may also need a special case for prereleases. We use the convention of adding "-prerelease" to the version number, which would make a prerelease always seem newer than the equivalent full release. Dev versions are kind of an edge case to begin with, though. Maybe this is good enough for main versions, and some followup work can address guiding people to the latest dev release if they're using a dev release.

IhateTrains commented on Fronter/Source/UpdateChecker/UpdateChecker.cpp:
```diff
@@ -0,0 +1,216 @@
+
+static int writer(char* data, size_t size, size_t nmemb, std::string* writerData)
+{
+	if (writerData == NULL)
```
Not sure if I can replace with `nullptr` as the NULL comes from CURL, but I will replace with `if (!writerData)`.

Idhrendur commented on Copy_Files.bat:
```diff
@@ -0,0 +1,18 @@
+echo on
```
Wouldn't just adding these files to the project in Visual Studio allow you to remove the triplicate copy of Copy_Files?

Zemurin replied:
I already did that. There's no copy_files.bat in current version. What you're looking at is some new code merged with all the reverts back to some ancient fronter version.

IhateTrains replied:
Solved ✅

Zemurin commented:
Please make a copy of your fork somewhere, make a clean sync with current, kill this PR, overwrite what's changed, and make a fresh PR so I can look at stuff that's actually changed. I removed those copy bats, for example, on Oct 7.

You reverted a lot of code, and you're reapplying a ton of existing changes.

IhateTrains commented:
> Please make a copy of your fork somewhere, make a clean sync with current, kill this PR, overwrite what's changed, and make a fresh PR so I can look at stuff that's actually changed. I removed those copy bats, for example, on Oct 7.
> 
> You reverted a lot of code, and you're reapplying a ton of existing changes.

No need to, I removed the `.bat`s and based the `.vcxproj` on upstream/master. I don't see code reversions, can you recheck the diff?
The change from `loglevel` to `logLevel` was intentional, because that didn't match camelCase.

Idhrendur commented on log.txt:
```diff
@@ -0,0 +1,2 @@
+2020-12-11 19:15:44  [WARNING] Configuration/fronter-configuration.txt not found!
```
log.txt shouldn't be checked in.
